src/components/section/helpers.ts

```diff
@@ -21,7 +21,9 @@ export function getFieldControlWithFallback(question: OHRIFormField) {
   if (question.type === 'encounterLocation') {
     question.questionOptions.rendering = 'encounter-location';
   }
-
+  if (question.type === 'encounterProvider') {
+    return getRegisteredControl('encounterProvider');
+  }
```

Comment:
@kajambiya, there are two problematic issues with this:

1. We are gradually coupling the datatype property with the rendering property.
2. This function is performing more work than intended. Initially, it was meant to serve as a fallback to text rendering for fields lacking concepts. However, it now involves some form of mapping based on the datatype. Ideally, the `type` prop should configure the handler component, while the `rendering` should configure the actual control (the underlying component). Over time, this function may become bloated and challenging to maintain. Frankly, I don't see a good reason for this approach.

Reply:
I understand the issue, @kajambiya. There are a couple of tweaks that could potentially solve it.

Let's lay all the pieces on the table: Our approach to supporting derived controls (components based on data sources plus the ui-select-extended component) relies on a combination of several elements:
 
-  **The ui-select-extended component (USE)**
 This is a generic component used to flesh out our derived control
- **A Datasource**
This is responsible for loading arbitrary data ie. coded answers from the backend 
- **A Submission handler**
This is responsible for form field value binding and a factory of composite field values

### How do we put these together?
At the heart of the form-engine is the registry, which manages all components of the form-engine. 
To get a derived control or rendering we [introduced control templates](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/control-templates.ts#L3) that provide the default DS config, which are then used to [derive new controls](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/inbuiltControls.ts#L120-L124). 

This approach isn't perfect but seems to work perfectly and keeps us operating within the premises of the Form Registry. 

### The simpler solution (Feels a bit hacky)
We don't have to worry about the rendering and submission handler because, at the end of the day, the correct components will be injected based on the config you shared above. However, we have to figure out how to inject the datasource along with the [default config](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/control-templates.ts#L13-L25). Currently, if an instance of a USE component has no datasource config as part of the form field configuration, [we fallback to using the field rendering type](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/components/inputs/ui-select-extended/ui-select-extended.tsx#L49-L53) to identify an existing inbuilt DS template. The magic bullet is falling back to the `question.type` as opposed to the `question.questionOptions.rendering`. We just have to make sure that the control template name aligns with the value of the type prop. 
NB: This is a potential breaking change for rendering types such as "drug" and "problem" but easy to fix. 

### Second solution (I seem to prefer this one)
We can simply introduce a form schema adaptor or something along those lines. Its primary role could be to ensure that the schema is fine-tuned to what the renderer expects. Before rendering the form, we could perform some kind of schema validation and refinement.

Input:

```json
{
  "type": "encounterLocation",
  "label": "Facility name (site/satellite clinic required):",
  "id": "location",
  "required": "true",
  "questionOptions": {
    "rendering": "ui-select-extended"
  }
}
```

Output:

```json
{
  "type": "encounterLocation",
  "label": "Facility name (site/satellite clinic required):",
  "id": "location",
  "required": "true",
  "questionOptions": {
    "rendering": "encounter-location"
  }
}
```
Here we simply change the rendering to [encounter-location](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/registry/inbuilt-components/inbuiltControls.ts#L60-L64) (Note that this rendering is not based on the USE) we are currently doing something similar [here](https://github.com/openmrs/openmrs-form-engine-lib/blob/8c5cf5ba6df5180af0eb2603bd0ce3202ae2b8ec/src/components/section/helpers.ts#L22-L23). Thoughts? cc: @ibacher @denniskigen

Reply:
Thanks @samuelmale 
I actually like both solutions. Let me wait to hear what other folks have to say